File: src/main/Triangle.java
import java.util.Arrays;
import java.util.List;

/**
 * 三角形最小路径和：https://leetcode-cn.com/problems/triangle/
 */
public class Triangle {

    public static int minimumTotal(List<List<Integer>> triangle) {
        int n = triangle.size();
        int[][] dp = new int[n][n];
        dp[0][0] = triangle.get(0).get(0);

        for (int i = 1; i < n; i++) {
            for (int j = 0; j <= i; j++) {
                if (j == 0) {
                    dp[i][j] = dp[i - 1][j] + triangle.get(i).get(j);
                } else if (j == i) {
                    dp[i][j] = dp[i - 1][j - 1] + triangle.get(i).get(j);
                } else {
                    dp[i][j] = Math.min(dp[i - 1][j - 1], dp[i - 1][j]) + triangle.get(i).get(j);
                }
            }
        }
        int result = dp[n - 1][0];
        for (int i = 1; i < n; i++) {
            result = Math.min(result, dp[n - 1][i]);
        }
        return result;
    }

    public static void main(String[] args) {
        System.out.println(minimumTotal(Arrays.asList(
            Arrays.asList(2),
            Arrays.asList(3,4),
            Arrays.asList(6,5,7),
            Arrays.asList(4,1,8,3)
        )));
    }

}

File: c/triangle.c
/**
 * 120. 三角形最小路径和：https://leetcode.cn/problems/triangle/
 */

#include <limits.h>
#include <string.h>

#define MIN(a,b) (((a)<(b))?(a):(b))

int dfs(int **triangle, int i, int j, int m, int memo[m][m]) {
    if (i == m) {
        return 0;
    }
    if (memo[i][j] != INT_MAX) {
        return memo[i][j];
    }
    return memo[i][j] = triangle[i][j] + MIN(dfs(triangle, i + 1, j, m, memo), dfs(triangle, i + 1, j + 1, m, memo));
}

int minimumTotal(int **triangle, int triangleSize, int *triangleColSize) {
    int memo[triangleSize][triangleSize];
    for (int i = 0; i < triangleSize; i++) {
        for (int j = 0; j < triangleSize; j++) {
            memo[i][j] = INT_MAX;
        }
    }
    return dfs(triangle, 0, 0, triangleSize, memo);
}

int minimumTotal2(int **triangle, int triangleSize, int *triangleColSize) {
    int f[triangleSize + 1][triangleSize + 1];
    memset(f, 0, sizeof(f));
    for (int i = triangleSize - 1; i >= 0; i--) {
        for (int j = 0; j <= i; j++) {
            f[i][j] = triangle[i][j] + MIN(f[i + 1][j], f[i + 1][j + 1]);
        }
    }
    return f[0][0];
}

int minimumTotal3(int **triangle, int triangleSize, int *triangleColSize) {
    int f[triangleSize + 1];
    memset(f, 0, sizeof(f));
    for (int i = triangleSize - 1; i >= 0; i--) {
        for (int j = 0; j <= i; j++) {
            f[j] = triangle[i][j] + MIN(f[j], f[j + 1]);
        }
    }
    return f[0];
}